### DISCUSSION TITLE
Add a module to send spam samples to vendors or other rspamd instances

### DISCUSSION THREAD

**Discussion by PHPGangsta** [rspamd/rspamd#4907] [Ideas]
Hi,

would it make sense to build a rspamd module to report spam samples to vendors?

I saw that Spamhaus (CC @pmcintyre ) is working on an API to report spam samples (and also single URLs, domains, IPs) via API:
https://submit.spamhaus.org/api

It would be awesome if rspamd could send spam samples to different anti-spam-vendors like Spamhaus, Abusix, rspamd.com or even other rspamd instances of my friends (using the "/learnspam" endpoint). Only if the administrator activates it and local law allows it of cause.

It would be awesome to have a "generic module" for this, so users can easily adapt/extend it to their needs. In most cases it's either a "forward the email to a specific email address" (if needed in XARF format) or a "HTTP call to a specific URL with specific headers and the raw email as body".

Some APIs (like rspamd) also accept a "JSON Control Body" instead of HTTP headers:
https://rspamd.com/doc/architecture/protocol.html#rspamd-json-control-block

The metadata-exporter of rspamd is meant to only send metadata to a HTTP service (that's the name of the module). With setting "meta_headers=true" it seems to also send the raw mail as HTTP body, but it looks like I cannot change/control the format of the headers to adapt it to the APIs of Spamhaus, Abusix or another rspamd-instance "/learnspam" endpoint.
https://rspamd.com/doc/modules/metadata_exporter.html

Possible vendors/targets to send spam samples to:

Spamhaus:
https://submit.spamhaus.org/api

Abusix:
https://docs.abusix.com/data-channels/5q9CV1FJbGR3vWZqVAYrqa/report-abusive-behaviour-to-abusix/mJc8WmBu78dsphxiweyDi6#forwarding-data-after-the-data-channel-is-configured

Another rspamd instance:
https://rspamd.com/doc/workers/controller.html
/learnspam

rspamd.com
Maybe rspamd.com also wants a copy of spam mails to improve its lists (fuzzy storage, or bl.rspamd.com come to my mind)?

**Comment by fatalbanana** [comment]
> would it make sense to build a rspamd module to report spam samples to vendors?

I think it would. Alternatively it might be part of `spamtrap` module; `metadata_exporter` is probably yet another option.

Abusix asks for some metadata which Rspamd doesn't have, unsure if they really require it or not, I'm so far sending them reports out of Rspamd with a little help from MTA.

**Reply by pmcintyre** [reply]
> would it make sense to build a rspamd module to report spam samples to vendors?

At spamhaus, we think this is a good idea, provided it can be limited to manually tagged spam samples (eg things submitted via /learnspam ).

If you start feeding automatically detected spam, and you have other (commercial) modules enabled, you may start submitting things that other companies detect, and it may lead to legal issues.

Spamhaus can help write the interface to our submission API in the community portal.

**Comment by fatalbanana** [comment]
Probably we should have one or more explicit interfaces & better not hide things in places like `/learnspam`.

Manual reporting might be implemented via an `rspamadm` command.

The idea of `spamtrap` module is that everything hitting some particular addresses is spam, so used properly it should be suitable for automatic feeds.

Reporting spam based on detections is definitely a poor idea.

We should probably take care to consider interests of potential recipients of reports.

EDIT: should have been a reply to [this thread](https://github.com/rspamd/rspamd/discussions/4907#discussioncomment-9048497) ...

**Comment by smfreegard** [comment]
I'm happy to get involved with this to enable reporting into the Abusix Global Reporting platform as well.   The benefit here is that you can (optionally) ask us to actually report the spam back to the responsible abuse desk automatically, so you have a greater chance of something actually getting done to prevent it from recurring.
